src/main/cpp/src/map_utils.cu

```diff
@@ -0,0 +1,633 @@
+
+  auto const joined_input = cudf::strings::detail::join_strings(
+      cudf::strings_column_view{input},
+      cudf::string_scalar(","),  // append `,` character between the input rows
+      cudf::string_scalar("{}"), // replacement for null rows
+      stream, rmm::mr::get_current_device_resource());
+  auto const joined_input_child =
+      joined_input->child(cudf::strings_column_view::chars_column_index);
+  auto const joined_input_size_bytes = joined_input_child.size();
+  CUDF_EXPECTS(joined_input_size_bytes + 2 == output_size, "Incorrect output size computation.");
+
+  // We want to concatenate 3 strings: "[" + joined_input + "]".
+  // For efficiency, let's use memcpy instead of `cudf::strings::detail::concatenate`.
+  auto output = rmm::device_uvector<char>(joined_input_size_bytes + 2, stream);
+  CUDF_CUDA_TRY(cudaMemsetAsync(output.data(), static_cast<int>('['), 1, stream.value()));
+  CUDF_CUDA_TRY(cudaMemcpyAsync(output.data() + 1, joined_input_child.view().data<char>(),
+                                joined_input_size_bytes, cudaMemcpyDefault, stream.value()));
+  CUDF_CUDA_TRY(cudaMemsetAsync(output.data() + joined_input_size_bytes + 1, static_cast<int>(']'),
+                                1, stream.value()));
+
+#ifdef DEBUG_FROM_JSON
+  print_debug<char, char>(output, "Processed json string", "", stream);
+#endif
+  return output;
+}
+
+// Check and throw exception if there is any parsing error.
+void throw_if_error(rmm::device_uvector<PdaTokenT> const &tokens,
+                    rmm::device_uvector<SymbolOffsetT> const &token_indices,
+                    rmm::cuda_stream_view stream) {
+  auto const error_count =
+      thrust::count(rmm::exec_policy(stream), tokens.begin(), tokens.end(), token_t::ErrorBegin);
+
+  if (error_count > 0) {
+    auto const error_location =
+        thrust::find(rmm::exec_policy(stream), tokens.begin(), tokens.end(), token_t::ErrorBegin);
+    SymbolOffsetT error_index;
+    CUDF_CUDA_TRY(cudaMemcpyAsync(
+        &error_index, token_indices.data() + thrust::distance(tokens.begin(), error_location),
+        sizeof(SymbolOffsetT), cudaMemcpyDeviceToHost, stream.value()));
+    stream.synchronize();
+    CUDF_FAIL("JSON Parser encountered an invalid format at location " +
+              std::to_string(error_index));
+  }
+}
+
+// Check if a token is a json node.
+struct is_node {
+  __host__ __device__ bool operator()(PdaTokenT const token) const {
+    switch (token) {
+      case token_t::StructBegin:
+      case token_t::ListBegin:
+      case token_t::StringBegin:
+      case token_t::ValueBegin:
+      case token_t::FieldNameBegin:
+      case token_t::ErrorBegin: return true;
+      default: return false;
+    };
+  }
+};
+
+// Compute the level of each token node.
+// The top json node (top json object level) has level 0.
+// Each row in the input column should have levels starting from 1.
+// This is copied from cudf's `json_tree.cu`.
```

Comment:
Did we need to copy this? Has it been changed?

Reply:
If I mark "This is copied" then there is no changes. These are helper functions that we need from cudf json code, but we can use them from cudf because it makes no sense to expose such minor helper functions into a public header.

Reply:
ok, so these are locked behind an anonymous or detail namespace, hence the copy. Ok.